### DISCUSSION TITLE
Managing implementation details for message data types

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#61] [Ideas]
As we progress with implementing import/exporting protobuf files we've come across some design questions regarding how we manage the complexity of implementation details with the abstract/generic definition of the AaC DSL. For instance, protobuf3 has an extensive list of supported data types [see the table here](https://developers.google.com/protocol-buffers/docs/proto3#scalar), but the AaC DSL currently only defines the following primitives (string, int, number, bool, date, file). As we map messages defined in AaC to these more complex protobuf implementations, we will be forced to either push extra complexity into the AaC definition or limit the protobuf3 implementation. 

In the current AaC language design what would the strategy be for allowing a system engineer to define a message without the complexity of understanding signed vs unsigned integers while not preventing the protobuf implementation from using signed vs unsigned integers?

This discussion is directed at generating data messages, but this same problem of moving from abstract -> concrete details will impact more than just the data message generation as we'll have the same problem when we're looking to generate other aspects of the architecture that are presented abstractly to the systems engineer who is modelling the system (e.g. infrastructure and service implementations).

**Comment by cvchaparro** [comment]
@jondavid-black

I want to add some context to Alex's comments above.

> As we map messages defined in AaC to these more complex protobuf implementations, we will be forced to either push extra complexity into the AaC definition or limit the protobuf3 implementation.

To add to this, we don't think these solutions are effective or exhaustive (see below). They aren't effective because if we update the AaC types, that is a bad precedent to set (eventually AaC would have the union of all the types for any message format we end up using). Additionally, limiting the protobuf3 implementation wouldn't be effective either because (again) it sets a precedent that we don't want to set and, furthermore, it ultimately limits Jellyfish 3.0 as a whole since it would mean we can only generate a subset of every type of messaging format that a user might want to use.

Another option I can think of (not listed above) is adding some kind of (possibly complex) system that allows sub-typing. For example, the AaC might have a `number` type for which an implementor might define a sub-type of `integer`, `float`, `uint64`, etc. Of course, this could quite possibly add more complexity than we want for an MVP (i.e. at this stage, implementing this would probably be over-engineering).

**Comment by Coffee2Bits** [comment]
I agree with Cameron that a path forward would be to allow users to incrementally transition from abstract definitions to more concrete definitions. The type/sub-type strategy could be effective, that's the initial idea I had for my jellyfish v3 DSL, but it certainly adds a lot of complexity to the DSL and project. 

We can alternatively just provide an extensive set of primitive data definitions to cover the general primitive data types that you'd find in programming languages (e.g. double, float, uint32, uint64, int32, int64, and bytes) and allow the team employing AaC to introduce the complexity themselves. For example, the systems engineer would create a message and use the generic type 'number' then another team member, such as a developer, would change the generic 'number' to a more concrete type 'uint64'. Of course, this method wouldn't hide the implementation details from the systems engineer who will now have to interact with 'uint64' every time they interact with our example message.

**Comment by jondavid-black** [comment]
The strategy we used for Jellyfish 2.0 was to "model the deployment" by extending the model with a set of attributes required to make the transition from abstract to concrete.  Perhaps we could consider something similar here.

Example:  https://github.com/NorthropGrumman/jellyfish/blob/master/jellyfish-examples/system-descriptor-projects/threat-eval-system-descriptor/src/main/sd/com/ngc/seaside/threateval/deployment/DemoThreatEvalDeployment.sd

I think this is an area where we should also consider a transition from one stakeholder type to another in the the workflow.  Someone interested in deciding between signed vs unsigned int is probably not someone dealing with top level system decomposition and use cases.  We certainly need to preserve the abstract.  Ideally we do that without driving too much complexity into the model specification itself.

Here's an initial thought...based off the JF experience.  Let's create a 'refine' root type that collects 'concrete implementation data' to be used by a tool performing the generation.  For a document, this may be something specifying an html template and css.  For Protobuf, this may be something that maps a AaC data type to a Protobuf data type.  Take a look at the extension root type and see if that concept may apply here. If we need, we could make this more complex by allowing nesting.

BTW...not a huge fan of the name 'refine', but don't have a better idea right now.

**Comment by jondavid-black** [comment]
Gents...please keep your eye on the project page in this repo.  I'm trying to keep track of things there in a kanban board.

This discussion pushed me to look at #21 and do some thinking.  I've realized that even though we're just getting started we already have some tech debt...mostly driven by my lack of python experience and the fact this started as a gradle driven java project before switching to python.

I think it is in our best interest to overcome that tech debt before we continue with this capability.  I've created a couple stories and put them at the top of the kanban to address that.  Given we're in the early stages, let's be sure to quickly identify and resolve tech debt...otherwise we'll struggle to meet our objectives in the time we've got.

**Comment by Coffee2Bits** [comment]
> The strategy we used for Jellyfish 2.0 was to "model the deployment" by extending the model with a set of attributes required to make the transition from abstract to concrete. Perhaps we could consider something similar here.
> 
> Example: https://github.com/NorthropGrumman/jellyfish/blob/master/jellyfish-examples/system-descriptor-projects/threat-eval-system-descriptor/src/main/sd/com/ngc/seaside/threateval/deployment/DemoThreatEvalDeployment.sd
> 
> I think this is an area where we should also consider a transition from one stakeholder type to another in the the workflow. Someone interested in deciding between signed vs unsigned int is probably not someone dealing with top level system decomposition and use cases. We certainly need to preserve the abstract. Ideally we do that without driving too much complexity into the model specification itself.
> 
> Here's an initial thought...based off the JF experience. Let's create a 'refine' root type that collects 'concrete implementation data' to be used by a tool performing the generation. For a document, this may be something specifying an html template and css. For Protobuf, this may be something that maps a AaC data type to a Protobuf data type. Take a look at the extension root type and see if that concept may apply here. If we need, we could make this more complex by allowing nesting.

I like the idea of the refine/deployment model that you linked from the JF repository--it reminds me of passing args to a terraform module to add extra context and implementation detail. That being said, if we used the `refine` properties to further define data types then there is a possibility that the refined data types will conflict with the abstract data types. We could just 'allow' this overriding behavior and treat the abstract message definition as the general structure of the data model while the refined properties as the source of truth for the data types. If we implemented this two-tiered definition for data models, and if we didn't implement any validation, then we would encounter some confusing/conflicting situations, like an abstract type `string` being overridden/implemented as a `uint32` type:

Message.yaml
```
data: 
  name: Message
  fields:
  - name: body
    type: string
  - name: sender
    type: string
 ---
 refine:
   target: Message
   properties:
   - fields.body.type: string
   - fields.sender.type: uint32
```

If we want to avoid the above situation we could leverage a type/subtype system for data types so that the modeler defines an abstract type, like `number` that would confine the developer's implementation type to a `number` sub-type like `int32`, `uint32`, `int64`, `uint64`, `float`, `double`. That being said, our first implementation can ignore validation of the abstract -> implementation data type.

I agree that `refine` doesn't sound like the best keyword. I'll throw `implementation` into the ring for consideration since the content under the `implementation` node will be just that: implementation details. It would look something like:

Message.yaml
```
data: 
  name: Message
  fields:
  - name: body
    type: string
  - name: sender
    type: string
  - name: message_id
    type: number
 ---
 implementation:
   implements: Message # Target any data or model definition by name
   properties:
   - fields.body.type: string
   - fields.sender.type: string
   - fields.message_id.type: uint32
```

**Comment by cvchaparro** [comment]
I recall the model/deployment structure we had in Jellyfish 2 (and, in fact, proposed a variation of that in my [example](https://gitlab.com/cvcf/cubozoa/-/blob/master/example/alarm/impl/configuration.lisp) that I called a `configuration`). In fact [here](https://gitlab.com/cvcf/cubozoa/-/blob/master/example/alarm/impl/configuration.lisp#L14-24) I was proposing something akin to what Alex proposes above.

I mention all of this to say that I like the idea of re-using the `deployment` and `refine` functionality from Jellyfish 2 and I would be happy to switch to using `implementation` (or `configuration`) over `refine` - whichever makes the most sense.

**Comment by Coffee2Bits** [comment]
I think for now we can get away with using extensions to add implementation details to models. I think the following would be the necessary steps to move forward.

- Define an `implementation` or `configuration` block (would be added to the model via extensions)
- Decide if implementation details should be an attribute override, or if they should be treated as additional context
  - Override - the property values defined in the `properties` attribute would replace the original values in the model
  - Additional context - the property values are associated with/mapped to the original fields/attributes and are available for features like generation, import/export, etc
- Decide how we handle property-override collisions in extensions.
  - Order-based? First one wins, last one wins?
- Decide if we want to limit what can be overridden (can implementation/configuration override/re-define `model` attributes or just `data` attributes?)
- Decide how we handle insufficient implementation details
  - Do we validate before generation?
  - Do we leverage default mappings so that an implementation can always be generated?
  - Do we fall back on runtime exceptions when presented with insufficient context/implementation details?

**Comment by jondavid-black** [comment]
Here's a wild idea.  Let's consider it from a validation perspective.  Right now the validator just checks structure, but maybe we need a way to allow plug-ins to add validation logic/rules when there's an extension applied.  Basically, we'll give the plugin developer structural validation for free, but also give them a hookspec where they can add custom validation.

So why did I go down this path?

I like the "additional context" approach you mention above, but I can see issues where you may need to check things like every field has an appropriately mapped "additional context" to pass validation.  I think it would be best if the user could get feedback via validation on their "implementation" as quickly as possible (i.e. missing field context for protobuf definition).  I think this would allow the plug-in developer some latitude in handling plugin specific collisions.  Cross-plugin conflicts is much harder...maybe not something we want to deal with right now.

I don't know if we should explicitly incorporate any limitations.  For example, I can imagine a model implementation that provides "additional context" for a particular behavior type (i.e. ports & protocols).

I think if we expand the plugin API to allow for custom validation the responsibility for handling insufficient implementation detail falls to the plugin developer.  I like the idea of default mappings, but again I think that is the plugin developer's responsibility (i.e. AaC int type may map to 32 bit signed integer by default).  Same philosophy for for the insufficient context/implementation details...it should be the responsibility of the plugin developer to determine what works best.

All this makes me realize we'll need some really robust documentation for plugin developers.

**Comment by Coffee2Bits** [comment]
I like the idea of letting plugin developers define what those implementation details look like for their plugin. That being said, that would require plugin developers to define their own data structure and extensions, and then users of the plugins will have to implement disparate configuration blocks for each of the plugins. I wonder if there is some common ground that we could find such as the standard DSL defining a configuration root key, but the onus of fleshing it out is on the plugin developer to define via extensions.

We have some larger/more difficult discussions if we want to have validator plugins (though I love the idea).

**Comment by Coffee2Bits** [comment]
The recent pull request #84 takes a pretty simple approach to this question, what if the plugin developer defined the implementation details they want?

I defined an enum with all of the protobuf types and extended the Field definition to support this new enum type.
```yaml
enum:
  name: ProtobufDataType
  values:
    - double
    - float
    - int32
    - int64
    - uint32
    - uint64
    - sint32
    - sint64
    - fixed32
    - fixed64
    - bool
    - string
    - bytes
---
ext:
   name: ProtobufTypeField
   type: Field
   dataExt:
      add:
        - name: protobuf_type
          type: ProtobufDataType
```

I think it worked fairly well and it's cool that we do this already, however, I don't think this is a good final solution. 

Pros:
- Already supported by the language
- Easy/intuitive to implement as someone familiar with the AaC DSL
- Locality is nice - my protobuf implementation type is next to the field it applies to

Cons:
- the enum attribute `protobuf_type` can be validly applied to ANY `Field` type instead of just the data messages that we're interested in
- A sufficiently complex modeled system with other plugins using the same pattern will create a lot of noise in the model
- There is no single source of truth - a repeated field is identified from the core `type`, but the actual type is sourced from `protobuf_type`

I think this pattern is adequate for now, but we'll want to keep thinking about a better solution.
